components/bpmn/org.wso2.carbon.bpmn.analytics.publisher/src/main/java/org/wso2/carbon/bpmn/analytics/publisher/utils/AnalyticsPublishServiceUtils.java

```diff
@@ -0,0 +1,301 @@
+			bpmnTaskInstance.setEndTime(taskInstance.getEndTime());
+			bpmnTaskInstance.setTaskName(taskInstance.getName());
+			bpmnTaskInstance.setDurationInMills(taskInstance.getDurationInMillis());
+			//bpmnTaskInstance.setWorkTimeInMills(taskInstance.getWorkTimeInMillis());
+			bpmnTaskInstance.setCreateTime(taskInstance.getCreateTime());
+			bpmnTaskInstance.setOwner(taskInstance.getOwner());
+			bpmnTaskInstance.setProcessInstanceId(taskInstance.getProcessInstanceId());
+			bpmnTaskInstances.add(bpmnTaskInstance);
+		}
+		return bpmnTaskInstances.toArray(new BPMNTaskInstance[bpmnTaskInstances.size()]);
+	}
+
+	/**
+	 * Format the process instance variables as BPMNVariable objects
+	 *
+	 * @param processVariables variables which belongs to a given process instance as key, value pairs
+	 * @return BPMNVariable objects array
+	 */
+	private BPMNVariable[] formatVariables(Map<String, Object> processVariables) {
+		if (processVariables == null) {
+			return null;
+		}
+		BPMNVariable[] vars = new BPMNVariable[processVariables.size()];
+		int currentVar = 0;
+		for (Map.Entry entry : processVariables.entrySet()) {
+			vars[currentVar] = new BPMNVariable(entry.getKey().toString(),
+			                                    processVariables.get(entry.getKey().toString())
+			                                                    .toString());
+			currentVar++;
+		}
+		return vars;
+	}
+
+	/**
+	 * Write last completed process instance end time to carbon registry
+	 *
+	 * @param historicProcessInstanceList List of historic process instances
+	 */
+	private void writePublishTimeToRegistry(
+			List<HistoricProcessInstance> historicProcessInstanceList) {
+		log.debug("Start writing last completed process instance publish time...");
+		Date lastProcessInstancePublishTime =
+				historicProcessInstanceList.get(historicProcessInstanceList.size() - 1)
+				                           .getStartTime();
+		try {
+			PrivilegedCarbonContext context = PrivilegedCarbonContext.getThreadLocalCarbonContext();
+			Registry registry = context.getRegistry(RegistryType.SYSTEM_GOVERNANCE);
+			Resource resource;
+			if (!registry.resourceExists(AnalyticsPublisherConstants.PROCESS_RESOURCE_PATH)) {
+				resource = registry.newResource();
+				resource.addProperty(AnalyticsPublisherConstants.LAST_PROCESS_INSTANCE_PUBLISH_TIME,
+				                     String.valueOf(lastProcessInstancePublishTime));
+				registry.put(AnalyticsPublisherConstants.PROCESS_RESOURCE_PATH, resource);
+			} else {
+				resource = registry.get(AnalyticsPublisherConstants.PROCESS_RESOURCE_PATH);
+				resource.setProperty(AnalyticsPublisherConstants.LAST_PROCESS_INSTANCE_PUBLISH_TIME,
+				                     String.valueOf(lastProcessInstancePublishTime));
+				registry.put(AnalyticsPublisherConstants.PROCESS_RESOURCE_PATH, resource);
+			}
+			log.debug("End of writing last completed process instance publish time...");
+		} catch (RegistryException e) {
```

Comment:
Add more information to the debug message. e.g. instance id, published time, etc